// lib/util/request-handler/base-request-handler.js
var streamCoupler = require('../stream-coupler');
var ViewStream = require('./view-stream.js');

function BaseRequestHandler() {

}

BaseRequestHandler.prototype.setRequest = function (request) {
  this._request = request;
};

BaseRequestHandler.prototype.setResponse = function (response) {
  this._response = response;
};

BaseRequestHandler.prototype.setErrorHandler = function (handler) {
  this._errorHandler = handler;
};

BaseRequestHandler.prototype.setRoute = function (route) {
  this._route = route;
};

BaseRequestHandler.prototype.execute = function (streams) {
  throw new Error('You must create your own execute method');
};

BaseRequestHandler.prototype._coupleErrorStreams = function (error) {
  streamCoupler(this._concatErrorStreams());
  this._viewStream.emit('streamError', error);
};

BaseRequestHandler.prototype._createViewStream = function () {
  this._viewStream = new ViewStream({
    errorHandler: this._errorHandler
  });
  return this;
};

BaseRequestHandler.prototype._concatStreams = function (streams) {
  throw new Error('You must create your own _concatStream method');
};

BaseRequestHandler.prototype._coupleStreams = function (streams) {
  streamCoupler(streams);
};

module.exports = BaseRequestHandler;
